src/pyFAI/ring_extraction.py

```diff
@@ -0,0 +1,388 @@
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+"""
+Module used for extracting control points in a calibrant image to be used for geometry refinement.
+"""
+
+from __future__ import annotations
+
+__authors__ = ["Emily Massahud", "Jérôme Kieffer"]
+__contact__ = "Jerome.Kieffer@ESRF.eu"
+__license__ = "MIT"
+__copyright__ = "European Synchrotron Radiation Facility, Grenoble, France"
+__date__ = "28/02/2024"
+__status__ = "development"
+
+
+import logging
+from concurrent.futures import ThreadPoolExecutor, as_completed
+from typing import Optional
+
+import numpy as np
+from silx.image import marchingsquares
+
+from .control_points import ControlPoints
+from .goniometer import SingleGeometry
+from .massif import Massif
+
+logger = logging.getLogger(__name__)
+
+
+class RingExtraction:
+    """Class to perform extraction of control points from a calibration image."""
+
+    def __init__(
+        self,
+        single_geometry: SingleGeometry,
+        massif: Optional[Massif] = None,
+    ):
+        """
+        Parameters
+        ----------
+        single_geometry : SingleGeometry
+            instance of pyFAI class
+        massif : Optional[Massif]
+            instance of Massif class, which defines an area around a peak; it is used to find
+            neighboring peaks, by default None
+        """
+
+        self.single_geometry = single_geometry
+        self.image = self.single_geometry.image
+        self.detector = self.single_geometry.detector
+        self.calibrant = self.single_geometry.calibrant
+
+        if massif:
+            self.massif = massif
+        else:
+            if self.detector:
+                mask = self.detector.dynamic_mask(self.image)
+            else:
+                mask = None
+            self.massif = Massif(self.image, mask)
+
+        self.two_theta_array = self.single_geometry.geometry_refinement.twoThetaArray()
+        self.two_theta_values = self._get_unique_two_theta_values_in_image()
+
+    def extract_control_points(
+        self,
+        max_number_of_rings: Optional[int] = None,
+        points_per_degree: int = 1,
```

Comment:
The number of points per degree is expected to be a floating point number (a value <1 is sometimes desired)